### PR TITLE
Ensure that PWM frequency is not changed by AUTO_FAN

### PR DESCRIPTION
**Pull request by andrei-moraru**
### Description

Fix `_UPDATE_AUTO_FAN` (in temperature.cpp) so it will not change timer PWM frequency during updates when a configuration with `EXTRUDER_AUTO_FAN_SPEED < 255` is set.

In the STM32 HAL, `analogWrite` will forcefully set the timer on every use to `1khz`. Causing `FAST_PWM_FAN_FREQUENCY` configuration to be overridden.

### Requirements

`AUTO_FAN` feature enabled with  `EXTRUDER_AUTO_FAN_SPEED < 255`.
Using `FAST_PWM_FAN` with a specific `FAST_PWM_FAN_FREQUENCY`.

### Benefits

`AUTO_FAN` feature will not interfere with `FAST_PWM_FAN_FREQUENCY` when  `EXTRUDER_AUTO_FAN_SPEED < 255`.

### Configurations

To test we need `AUTO_FAN_PIN` is enabled and `EXTRUDER_AUTO_FAN_SPEED < 255`.

This was tested on a `BIGTREE_SKR_2` (STM32 HAL). Measuring the frequency with an oscilloscope verified the fix.
- Before: It was always 1khz
- After: It follows `FAST_PWM_FAN_FREQUENCY`

### Related Issues

None

### CODE REVIEW

**Comment by thinkyhead** [comment]
This is incomplete because `set_pwm_duty` is not implemented in Marlin for all platforms. To get the PR finished we will need implementations of `set_pwm_duty` for DUE, ESP32, Linux, SAMD51, and Teensy.

**Comment by andrei-moraru** [comment]
Thanks, defining a HAL for `set_pwm_duty` makes more sense. Didn't realize its not defined everywhere. Thanks a lot for fixing it elsewhere. Couldn't figure out the how to run the test-suite locally, new to the whole embedded ecosystem of tools.

Makes sense to me to try and expand it beyond the `AUTO_FAN` feature since they all seem to be PWM related adjustments.

I don't know how to figure out if preferring `analogWrite` is better if `NEEDS_HARDWARE_PWM` is not defined over always exposing `set_pwm_frequency`.
- I would prefer to use the same implementation of `set_pwm_duty` for both cases (`NEEDS_HARDWARE_PWM` used or not) since there is less deviation between a user who enables the flag vs another who doesn't. I just don't know enough about all the hardware and its capabilities to know if that idea is valid and for what platforms. (Maybe 32-bit we can assume PWM is always-supported)

For example:
Looking at STM32 implementation of `analogWrite` it seems to support both `analog (DAC)`, `pwm (TIMER)` if not it defaults to `digital (HIGH/LOW only)`. 
- But for driving a FAN/Controlling a stepper etc.., would `analog` ever be used? (Can a FAN even be driven off the analog pin (Power draw) 
- For the digitalWrite case Marlin seems to handle that by itself (If 255 => call digitalWrite) thus we only use it for the PWM case and can always safely default to `set_pwm_duty` in theory.

**Comment by thinkyhead** [comment]
The use of the word "analog" in the context of writing is understandably confusing. When we do `analogRead` we're initiating the DAC, waiting until it has enough time to accumulate a value, and then we're reading the ADC result. But `analogWrite` has no relationship to ADC. Instead we're setting a PWM duty cycle, and the device receiving power is getting some fraction of the total 5V (or 12V/24V for heaters) based on the DC. (A fan would not be directly attached to a 3.3V pin but would have a transistor and 5V from elsewhere.)

**Comment by thinkyhead** [comment]
I'm not entirely sure why `digitalWrite` is used over `analogWrite` for a value of 0 or 255, except that it simplifies the pin logic so the frequency-generator and dividers could potentially be used for something else.

**Comment by tpruvot** [comment]
from what i see in framework code, analogWrite also set the port as output if not made before... default might be an input

cores/arduino/wiring_analog.c:200
```
      if (pin_in_pinmap(p, PinMap_TIM)) {
        if (is_pin_configured(p, g_anOutputPinConfigured) == false) {
          set_pin_configured(p, g_anOutputPinConfigured);
        }
        ulValue = mapResolution(ulValue, _writeResolution, _internalWriteResolution);
        pwm_start(p, _writeFreq, ulValue, _internalWriteResolution);
      }
```

edit: in fact its not that, set_pin_configured is just a bit set which means pwm_start was called...

#define set_pin_configured(pin, map) (map[PINCONF_INDEX(pin)] = map[PINCONF_INDEX(pin)] | PINCONF_BIT(pin))

File: Marlin/src/HAL/STM32F1/fast_pwm.cpp [hunk 1]
```diff
@@ -23,46 +23,46 @@
...
-    prescaler = 65535;
-    preload = (HAL_TIMER_RATE) / (prescaler + 1) / f_desired - 1;
-  }
-  if (prescaler < 0) return;                    // Too high frequency
-  timer_set_reload(timer, preload);
-  timer_set_prescaler(timer, prescaler);
-}
-
 void set_pwm_duty(const pin_t pin, const uint16_t v, const uint16_t v_size/*=255*/, const bool invert/*=false*/) {
   timer_dev *timer = PIN_MAP[pin].timer_device;
```
Comment (tpruvot) [inline Marlin/src/HAL/STM32F1/fast_pwm.cpp:31]: this code is not safe imo... `if (!PWM_PIN(pin)) return;` missing

Comment (descipher) [reply Marlin/src/HAL/STM32F1/fast_pwm.cpp:31]: Agreed! IIRC it is not available in the LPC library.

**Comment by descipher** [comment]
To use set_pwm_duty instead of analogWrite() I believe you will need to call pin_mode() and set the pin as output if it was not set previously.

File: Marlin/src/HAL/STM32F1/fast_pwm.cpp (hunk 1 above)
Comment (descipher) [reply Marlin/src/HAL/STM32F1/fast_pwm.cpp:31]: lol 
LPC HAL:
#define PWM_PIN(P)            true // all pins are PWM capable

**Comment by thinkyhead** [comment]
Pins should already be initialized as OUTPUT or as PWM (depending on platform) if they will be used for PWM. If some pin is being used with `set_pwm_duty` but is not already initialized, then its initialization will need to be added.

**Comment by roffe** [comment]
This seems to have broken parts cooling on MKS Robin nano 1.2 (PB1), if I build from commit before this was merged parts cooling works again

**Comment by tpruvot** [comment]
maybe try to add in board pins.h #define BOARD_INIT() analogWrite(FAN_PIN, 0) 

From what i see in robin nano variant (MARLIN_F103Vx), its TIMER1 but PB1 can in theory use 3 different timers

maybe BOARD_PREINIT is required to allow marlin init set the right pwm frequency then

**Comment by stuartpittaway** [comment]
This has broken the builds I've been using for the Index Pick & Place project.

The changes introduced in this commit cause problems with the GPIO outputs going high on power up, which work fine in previous versions (up to this commit 767a15d468ab6859eec19600541ad21dff757d1a).

The pick and place uses the FAN outputs to drive pumps and solenoid valves rather than a fan, so PWM isn't needed for this purpose.

**Comment by descipher** [comment]
> This has broken the builds I've been using for the Index Pick & Place project.
> 
> The changes introduced in this commit cause problems with the GPIO outputs going high on power up, which work fine in previous versions (up to this commit [767a15d](https://github.com/MarlinFirmware/Marlin/commit/767a15d468ab6859eec19600541ad21dff757d1a)).
> 
> The pick and place uses the FAN outputs to drive pumps and solenoid valves rather than a fan, so PWM isn't needed for this purpose.

Configuring #define FAN_SOFT_PWM could be your solution. It uses no PWM calls.

**Comment by stuartpittaway** [comment]
> Configuring #define FAN_SOFT_PWM could be your solution. It uses no PWM calls.

Thanks for the assistance, that does appear to help resolve the issue.

**Pull request merged**